### PR TITLE
feat: remove redudant approvals for certain DEXes

### PR DESCRIPTION
**Pull request by Wertusser**

### CODE REVIEW

File: src/executor/Executor01BytecodeBuilder.ts [hunk 1]
```diff
@@ -268,6 +268,7 @@ export class Executor01BytecodeBuilder extends ExecutorBytecodeBuilder<
       (!isETHAddress(swap.srcToken) ||
         (isETHAddress(swap.srcToken) && index !== 0)) &&
       !curExchangeParam.transferSrcTokenBeforeSwap &&
+      !curExchangeParam.skipApproval &&
```
Comment (KanievskyiDanylo) [inline src/executor/Executor01BytecodeBuilder.ts:271]: Assume it also should be included on line 294

File: src/executor/Executor02BytecodeBuilder.ts [hunk 2]
```diff
@@ -606,6 +606,7 @@ export class Executor02BytecodeBuilder extends ExecutorBytecodeBuilder<
     if (
       !isETHAddress(swap.srcToken) &&
       !curExchangeParam.transferSrcTokenBeforeSwap &&
+      !curExchangeParam.skipApproval &&
```
Comment (KanievskyiDanylo) [inline src/executor/Executor02BytecodeBuilder.ts:609]: Should also be included on line 625

File: src/executor/Executor03BytecodeBuilder.ts [hunk 3]
```diff
@@ -147,6 +147,7 @@ export class Executor03BytecodeBuilder extends ExecutorBytecodeBuilder<
     if (
       flags.dexes[index] % 4 !== 1 && // not sendEth
       !isETHAddress(swap.srcToken) &&
+      !curExchangeParam.skipApproval &&
```
Comment (KanievskyiDanylo) [inline src/executor/Executor03BytecodeBuilder.ts:150]: same here for the next approval before deposit

File: src/generic-swap-transaction-builder.ts [hunk 4]
```diff
@@ -674,6 +674,7 @@ export class GenericSwapTransactionBuilder {
     priceRoute: OptimalRate,
     dexExchangeParams: DexExchangeParam[],
     maybeWethCallData?: DepositWithdrawReturn,
+    skipApproval?: boolean,
```
Comment (KanievskyiDanylo) [inline src/generic-swap-transaction-builder.ts:677]: This param is never passed, looks redundant

File: src/dex/simple-exchange.ts [hunk 5]
```diff
@@ -110,12 +110,12 @@ export class SimpleExchange {
...
+    skipAllowance?: boolean,
   ): Promise<SimpleExchangeParam> {
-    const approveParam = await this.getApproveSimpleParam(
-      src,
-      spender || swapCallee,
-      srcAmount,
-    );
+    const approveParam = await (!skipAllowance
+      ? this.getApproveSimpleParam(src, spender || swapCallee, srcAmount)
+      : { callees: [], calldata: [], values: [], networkFee: '0' });
```
Comment (KanievskyiDanylo) [inline src/dex/simple-exchange.ts:117]: imho this looks cleaner
```suggestion
   const approveParam = skipAllowance
      ? { callees: [], calldata: [], values: [], networkFee: '0' }
      : await this.getApproveSimpleParam(src, spender || swapCallee, srcAmount);
```

File: src/dex/spark/spark.ts [hunk 6]
```diff
@@ -207,6 +207,10 @@ export class Spark
       destAmount,
       swapData,
       exchange,
+      undefined,
+      undefined,
+      undefined,
+      isSell && this.isDai(destToken),
```
Comment (KanievskyiDanylo) [inline src/dex/spark/spark.ts:213]: `skipApproval` is missing in getDexParam?

File: src/executor/Executor03BytecodeBuilder.ts (hunk 3 above)
Comment (Wertusser) [reply src/executor/Executor03BytecodeBuilder.ts:150]: fixed

File: src/executor/Executor02BytecodeBuilder.ts (hunk 2 above)
Comment (Wertusser) [reply src/executor/Executor02BytecodeBuilder.ts:609]: fixed

File: src/executor/Executor01BytecodeBuilder.ts (hunk 1 above)
Comment (Wertusser) [reply src/executor/Executor01BytecodeBuilder.ts:271]: fixed

File: src/dex/spark/spark.ts (hunk 6 above)
Comment (Wertusser) [reply src/dex/spark/spark.ts:213]: fixed

File: src/generic-swap-transaction-builder.ts (hunk 4 above)
Comment (Wertusser) [reply src/generic-swap-transaction-builder.ts]: fixed

File: src/dex/simple-exchange.ts (hunk 5 above)
Comment (Wertusser) [reply src/dex/simple-exchange.ts]: agree, fixed

File: src/dex/simple-exchange.ts [hunk 7]
```diff
@@ -110,12 +110,12 @@ export class SimpleExchange {
     spender?: Address,
     networkFee: NumberAsString = '0',
     preCalls?: Omit<SimpleExchangeParam, 'networkFee'>,
+    skipAllowance?: boolean,
```
Comment (0xNazarii) [inline src/dex/simple-exchange.ts:113]: lets name this param `skipApprove` too

**Review by 0xNazarii — APPROVED** [review]

File: src/dex/simple-exchange.ts (hunk 7 above)
Comment (Wertusser) [reply src/dex/simple-exchange.ts]: fixed

**Pull request merged**